ngraph/core/src/op/topk.cpp

```diff
@@ -508,7 +508,11 @@ void op::v1::TopK::validate_and_infer_types()
             auto max_k = maximum_value(input_value(1));
             if (max_k.first)
             {
-                output_shape[m_normalized_axis] &= Dimension(0, max_k.second);
+                const auto in_min = input_partial_shape[m_normalized_axis].get_min_length();
+                const auto in_max = input_partial_shape[m_normalized_axis].get_max_length();
+                const auto lower = std::min<Dimension::value_type>(in_min, max_k.second);
+                const auto upper = std::max<Dimension::value_type>(in_max, max_k.second);
```

Comment:
Just be aware that [Dimension::get_max_length()](https://github.com/openvinotoolkit/openvino/blob/9c509e5f416cce29dddd3bd1de292055df3765c5/ngraph/core/src/dimension.cpp#L140) will return `-1` for the upper bound limit `Interval::s_max`